### PR TITLE
Jsoo 406

### PR DESCRIPTION
**Pull request by ryyppy**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -178,19 +179,125 @@ This is usually the file you want to create to test certain compile behavior
...
+The "BuckleScript Playground bundle" is the BS compiler compiled to JavaScript, including all necessary dependency files (stdlib / belt etc).
+It is useful for building tools where you want to compile and execute arbitrary Reason / OCaml in the browser.
+
+The BuckleScript source code is compiled with a tool called [JSOO (js_of_ocaml)](https://ocsigen.org/js_of_ocaml/3.5.1/manual/overview), which uses OCaml
+bytecode to compile to JavaScript and is part of the bigger OCaml ecosystem. Before we can compile anything, we need to install the required
+tools (requires [`opam`](https://opam.ocaml.org/doc/Install.html) to be installed):
+
+```
+# Makes sure to be on the right switch
+opam switch 4.06.1
```
Comment (thangngoc89) [inline CONTRIBUTING.md:210]: `opam switch create 4.06.1` if you're doing this for the first time

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -178,19 +179,125 @@ This is usually the file you want to create to test certain compile behavior
...
+eval `opam config env`
+
+opam install js_of_ocaml.3.4.0
```
Comment (thangngoc89) [inline CONTRIBUTING.md]: should be at least 3.5.1

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -178,19 +179,125 @@ This is usually the file you want to create to test certain compile behavior
...
+```
+
+### Building the bundle
+
+The entry point of the JSOO bundle is located in `jscomp/main/jsoo_main.ml` and the script for running JSOO can be found in `scripts/repl.js`.
+A full clean build can be done like this:
+
+```
+# We create a target directory for storing the bundle / stdlib files
+mkdir playground && mkdir playground/stdlib
```
Comment (thangngoc89) [inline CONTRIBUTING.md:223]: `mkdir -p playground/stdlib`would work

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -178,19 +179,125 @@ This is usually the file you want to create to test certain compile behavior
...
+# Now we run the repl.js script pointing to our playground directory (note how it needs to be relative to the repl.js file)
+BS_PLAYGROUND=../playground node scripts/repl.js
+```
+
+**You should now find following files:**
+
+- `playground/exports.js` -> This is the BuckleScript compiler, which binds the BuckleScript API to the `window` object
+- `playground/stdlib/*.js` -> All the BuckleScript runtime files
+
+You can now use the `exports.js` file either directly by using a `<script src="/path/to/exports.js"/>` inside a html file, or use a bundler infrastructure to optimize it.
```
Comment (thangngoc89) [inline CONTRIBUTING.md]: or you can use it in Node.js environment like this:

```sh
$ node
> require("./exports.js");
undefined
> let compile_result = ocaml.compile(`Js.log Sys.ocaml_version`); // You can change the code here
undefined
> eval(compile_result);
4.06.2+BS
undefined
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -178,19 +179,125 @@ This is usually the file you want to create to test certain compile behavior
...
+target, where BuckleScript doesn't have access to the filesystem. When working
+on BS, you'll see diffs on those files whenever there are changes on core
+modules, e.g. stdlib modules or when the ocaml version was changed. We usually
+check in these files to keep it in sync with the most recent compiler
+implementation. JSOO will pick up those files to encode them into the `exports.js`
+bundle.
+
+For any other dependency needed in the playground, such as `ReasonReact`, you
+will be required to serialize your `.cmi` / `.cmt` files accordingly so that
+JSOO's `ocaml.load` function can load the data. Right now we don't provide any
```
Comment (thangngoc89) [inline CONTRIBUTING.md]: `ocaml.load_module` , but this has nothing to do with `JSOO`. It's a function defined in `jsoo_main.ml`

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -178,19 +179,125 @@ This is usually the file you want to create to test certain compile behavior
...
+`jscomp/core/js_cmi_datasets.ml`. These files are only linked for the browser
+target, where BuckleScript doesn't have access to the filesystem. When working
+on BS, you'll see diffs on those files whenever there are changes on core
+modules, e.g. stdlib modules or when the ocaml version was changed. We usually
+check in these files to keep it in sync with the most recent compiler
+implementation. JSOO will pick up those files to encode them into the `exports.js`
+bundle.
+
+For any other dependency needed in the playground, such as `ReasonReact`, you
+will be required to serialize your `.cmi` / `.cmt` files accordingly so that
```
Comment (thangngoc89) [inline CONTRIBUTING.md]: serialized the binary format into hex encoded string

**Review by thangngoc89 — CHANGES_REQUESTED** [review]
I left some comments

File: CONTRIBUTING.md (hunk 3 above)
Comment (ryyppy) [reply CONTRIBUTING.md:223]: i kept it simple since not all platforms support `-p` afaik

File: CONTRIBUTING.md (hunk 4 above)
Comment (ryyppy) [reply CONTRIBUTING.md]: cool, I didn't know that jsoo accounts for node as well!

File: CONTRIBUTING.md (hunk 5 above)
Comment (ryyppy) [reply CONTRIBUTING.md]: good catch

**Comment by ryyppy** [comment]
@thangngoc89 updated!

File: CONTRIBUTING.md (hunk 4 above)
Comment (thangngoc89) [reply CONTRIBUTING.md]: I actually have a simple assert testing for exports.js at some point. We can add it back :)

**Review by thangngoc89 — APPROVED** [review]

**Pull request closed without merging**